Feature/#66 pan while zoom

markusressel opened the pull request:
fixes #66 

Wow so this took quite a while for me to get it to work but I'm quite happy with the results.
The hardest part was animating back to a non-overscrolled and non-overpinched state at the same time.
I use a "simulated" zoom state to calculate the pan fixes needed to be applied to get back into a valid state when the zoom is also changing. I will add a review comment to the mentioned line, please let me know if you know a better way to do this.

markusressel committed:
apply pan based on the focus point change

markusressel committed:
fix overscroll animation on pinch end

markusressel committed:
documentation fixes
use Float.NaN instead of float comparison using epsilon
always animate when overscroll/overpinch is enabled since newZoom will always be > 0

markusressel committed:
allow overpinch and overscroll simultaniously in applyZoomAndAbsolutePan method (still a bug when animating back from an overpinched state)

markusressel committed:
removed unnecessary float/int conversion

markusressel committed:
changed pivot point of applyZoomAndAbsolutePan

markusressel committed:
use coerceIn
added fields for initial and current target of pinch gesture
improved zoom target coordinate selection based on which side of the view has been overscrolled during pinch gesture
only fix overscroll and/or overpinch if necessary
parameter renaming
some code cleanup

markusressel committed:
added documentation

markusressel committed:
use zoom simulation to calculate correct panFix values
extracted pivot point calculation to method

markusressel committed:
added commentary to describe what is happening
slight optimization

markusressel committed:
revert default animation duration

markusressel commented on library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:
```diff
@@ -753,54 +764,154 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
+                        mCurrentAbsTargetX, mCurrentAbsTargetY,
+                        allowOverScroll = true,
+                        allowOverPinch = true)
                 return true
```
This is the stuff (above code) that allows pan and zoom to be applied simultaniously when using pinch gesture.

markusressel commented on library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:
```diff
@@ -753,54 +764,154 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
+                    var newPanY = panY + fixPanY
+
+                    if (newZoom.compareTo(zoom) != 0) {
+                        // we have overpinched. to calculate how much pan needs to be applied
```
Here is the zoom "simulation" part. The problem I faced was that I couldn't just calculate pan fix with the current zoom and apply it at the same time as the new zoom level as the pan fix calculated by the `checkPanBounds` method is an absolute pan value that is always located at the top left of the view (I think?). Therefore this absolute position depends on the current zoom level - so the calculated fix is not applicable to the state after the overpinched zoom has been changed.

natario1 replied:
I should take a better look to understand what's happening. Anyway I just tested the app and it's not super smooth so there must be a bug somewhere. 

For instance if I pinch out (move fingers closer) then release, it suddenly moves to top-left then goes back. It seems to do this during pinch+pan also. Do you have any idea?

markusressel replied:
You are right when zooming out more than the size of the content (so there are gray areas all around the image) the view does jump when pinching out.
I will have a look at that.

I don't know of any other jumping though if this is not the scenario you are describing.

markusressel replied:
I think I fixed it using the center of `mContentRect` as the pivot point when zooming in from underpinch. @natario1 Could you have another look?

natario1 replied:
Seems to work great now! Nice work!

markusressel commented on library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:
```diff
@@ -753,54 +764,154 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
+                        newPanY = panY + fixPanY
+
+                        // revert simulation
+                        applyZoom(oldZoom, true, true, zoomTargetX, zoomTargetY)
```
Those `applyZoom` calls currently trigger listeners. I don't know if that's a problem but I think it should be avoided as it's just a simulation to calculate pan fix numbers?

markusressel replied:
I added a boolean parameter `triggerListeners` to the relevant methods to prevent triggering them when changes are only simulated.

markusressel commented on library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:
```diff
@@ -1240,29 +1364,30 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
+    private fun applyPinch(@Zoom zoom: Float, @AbsolutePan targetX: Float, @AbsolutePan targetY: Float,
                            allowOverPinch: Boolean) {
-        var newZoom = newZoom
+        // TODO: this method is currently unused, should it be removed? The doc mentions buggy translations :S
```
You used this method to apply zoom **only** in the `PinchListener`.
I don't know if it is of any other value and should be kept as some kind of helper?

natario1 replied:
If it's unused we can remove it.

markusressel replied:
Ok

markusressel commented on library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:
```diff
@@ -1044,25 +1155,23 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
             val startTime = System.currentTimeMillis()
-            @Zoom val startZoom = zoom
-            @Zoom val endZoom = newZoom
+            @Zoom val startZoom = this.zoom
```
I'm sorry - there are a number of small refactorings and variable renamings in this PR.
Android Studio was getting on my nerves with duplicate variable names (parameter and local variable with the same name) and I had a hard time understanding what the `time` variable actually is used for so I changed those things.
If there is anything that speaks against this I can try to reverse those changes - but if you want this please let me know exactly what I need to reverse.

natario1 replied:
I think it's fine

markusressel commented on library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:
```diff
@@ -753,54 +764,154 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
+
+                // apply the new zoom and pan values accordingly
+                applyZoomAndAbsolutePan(newZoom,
+                        mCurrentAbsTargetX, mCurrentAbsTargetY,
```
I just realized this pinch is missing the zoom target

markusressel replied:
fixed

markusressel commented on library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:
```diff
@@ -753,54 +764,154 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
+                scaledFocusY += scaledPanY
+
+                mCurrentAbsTargetX = panX
+                mCurrentAbsTargetY = panY
```
~This should be nInitialAbsTargetX if I'm not mistaking. this might cause the little jump at the beginning when pinching fast.~

markusressel replied:
Forget this, I was mistaking.
I though about this pan shifting while pinching again and it's now fixed.
Its much more smooth, no initial jump and works in all situations I could come up with.

markusressel committed:
refactored fields to use PointF instead of raw Float values
extracted viewCoordinateToAbsolutePan to method
fixed initial jump when zooming in fast
fixed strange wobble when zooming in and out fast

markusressel committed:
added AbsolutePoint and ScaledPoint data classes for easier handling of X and Y values at the same time
added simple JUnit5 test
added helper/extension functions

markusressel committed:
added methods to ScaledPoint too

markusressel committed:
keep the original zoomTarget even when an additional view boundary will be overscrolled due to overpinch adjustments (doesn't jump on corners)

markusressel committed:
changed comment to be more precise

markusressel committed:
var can be val

natario1 commented on library/src/main/java/com/otaliastudios/zoom/AbsolutePoint.kt:
```diff
@@ -0,0 +1,49 @@
+/**
+ * This class represents an absolute point on the ZoomEngine canvas (or beyond it's bounds)
+ */
+data class AbsolutePoint(
```
These make everything cleaner, thanks.
Are they exposed? If not let's make them internal.

markusressel replied:
As `panX` and `panY` are exposed I also exposed `pan` which is an `AbsolutePoint`. I can. remove that if it should only be an internal representation but I think the easier handling of points can also be beneficial to devs using this library - don't you think?

natario1 replied:
Yes, cool!

natario1 commented on library/src/test/java/com/otaliastudios/zoom/PointTests.kt:
```diff
@@ -0,0 +1,43 @@
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.TestInstance
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
```
Thanks for tests!

markusressel replied:
Well they are pretty basic so far, I just wanted to get the basic structure working. 

Initially I wanted to test methods of the engine but I failed to create a test that has a valid context that is needed by the engine :-(
I think this can be another PR though.

natario1 commented on library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:
```diff
@@ -1235,34 +1356,81 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
+    /**
+     * Converts an [AbsolutePoint] to a [ScaledPoint]
+     */
+    private fun AbsolutePoint.toScaled(): ScaledPoint {
```
Nice. I think we could finally remove `resolvePan` and `unresolvePan` and create extensions of the `Float` class with the same names?

natario1 replied:
I mean to call them toScaled() and toAbsolute() of course.

markusressel replied:
Sure.

markusressel committed:
Merge remote-tracking branch 'origin/master' into feature/#66_pan_while_zoom

markusressel committed:
added default values to AbsolutePoint and ScaledPoint constructor
added unaryMinus operator
replaced resolvePan and unresolvePan methods with kotlin extension functions
added separate tests for absolutePoint and ScaledPoint

markusressel committed:
replaced individual variables for x and y values with AbsolutePoint

markusressel committed:
added "*" operator to AbsolutePoint and ScaledPoint
use Number as input parameter for operators so Int and Float can be used
make more use of AbsolutePoint instead of two variables

markusressel committed:
make more use of ScaledPoint instead of two variables

markusressel committed:
added target annotation to RealZoom, Zoom, AbsolutePan, ScaledPan and ZoomType

markusressel committed:
define archive basename (because "app" is kind of useless)

markusressel committed:
use the center of the mContentRect as a pivot point when scaling back from underpinch (zoom value lower than minZoom)

markusressel committed:
also reset pan after simulation
make more use of ScaledPoint instead of two variables

markusressel committed:
added method to simplify "pan fix" calculation
more ScaledPan usage

markusressel committed:
renamed resetState() method in PinchListener and added documentation

markusressel committed:
sind androidx support lib version numbers are not in perfect sync anymore so a variable for those dependencies doesn't make sense
added named parameters to many boolean parameters (to make easier to understand without IDE support like on GitHub)

markusressel committed:
added methods for minZoom and maxZoom
replaced resolveZoom method with Float extension function

markusressel committed:
cancel any currently running animation when touching the screen - fixes #61
renamed function to match its usecase
use androids ValueAnimator for animations instead of pure Runnables and manual interpolation

markusressel committed:
use PropertyValuesHolder to animate multiple animators concurrently

markusressel committed:
ignore lint issue

markusressel committed:
doku fix

natario1 commented on app/build.gradle:
```diff
@@ -10,6 +10,8 @@ android {
         versionCode 1
         versionName "1.0"
+
+        setProperty("archivesBaseName", "ZoomLayout_v${versionName}_(${versionCode})")
```
What will this do?

markusressel replied:
When building an apk (release or debug) the resulting apk file will be called this way.
I have too much `app-debug.apk` files in my life already :stuck_out_tongue:

natario1 commented on library/src/main/java/com/otaliastudios/zoom/ZoomApi.kt:
```diff
@@ -88,6 +91,7 @@ interface ZoomApi {
      */
+    @Target(AnnotationTarget.FIELD, AnnotationTarget.FUNCTION, AnnotationTarget.LOCAL_VARIABLE, AnnotationTarget.PROPERTY, AnnotationTarget.VALUE_PARAMETER)
     @Retention(AnnotationRetention.SOURCE)
     annotation class ScaledPan
```
I regret using these name, something like AbsoluteCoord would have been better. But we will change in another version

markusressel replied:
Those names still confuse me :grin: 
They are also kind of useless since the IDE seems to do nothing to check these.
Even when passing wrong values to a function with annotated parameters there is no warning.
That's another reason why I opted to create data objects for them (although you can still pass in wrong values on their constructors...)

natario1 replied:
Ah, without them it would be worse I think... There should be an android annotation to tell the IDE that these two annotations exclude each other, but I don't know how.

markusressel replied:
Yes they do help indeed! But not to the extend I'd like them to :stuck_out_tongue:

natario1 commented on library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:
```diff
@@ -150,9 +150,14 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
+     * The current pan as an [AbsolutePoint]
+     */
+    private val pan: AbsolutePoint
+        get() = AbsolutePoint(panX, panY)
```
Are these pan and scaledPan values accessed frequently?
If they are it might be worth to avoid all these new instances by caching the value (have a private _pan and use _pan.set() in this getter), since touch events are a lot.

By the way I thought from another comment that these two were public?

markusressel replied:
I made them public now, don't know why they weren't.
Yes using the same Object is probably better for performance - I changed that.

natario1 replied:
I didn't know you could do what you did :-) just one last change, can they be vals of ZoomApi overriden in the engine?

markusressel replied:
Well the `pan` val can, the `scaledPan` depends on `mTransformedRect` so I don't think so?

natario1 commented on library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:
```diff
@@ -625,10 +637,25 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
+     * @return the scaled pan fix values for x- and y-axis
+     */
+    private fun getPanCorrection(): ScaledPoint {
+        return ScaledPoint(
```
See if it's better to avoid this new instance as well depending on how frequently this is accessed

markusressel replied:
done

natario1 commented on library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:
```diff
@@ -773,58 +806,171 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
 
         override fun onDown(e: MotionEvent): Boolean {
+            // cancel any animation that might still be running
+            setState(NONE)
```
I see some issues with this: zoom in, then fling fast toward one end, then put your finger when the content is overscrolled. It does not always come back.

Might be easy to fix, but we can leave it for another PR I think.

markusressel replied:
ok I will just revert this little change in this PR then.

natario1 commented on library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:
```diff
@@ -1056,86 +1202,118 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
 
     //region Apply values
 
+    private val mCancelAnimationListener = object : Animator.AnimatorListener {
```
Do you think we should setState(NONE) onAnimationCancel also?

markusressel replied:
Well the animation is only cancelled when `mClearAnimation` is set to `true` which is done in `setState(NONE)` ... so currently it's the other way around - the state change causes the animationListener to trigger onCancel. It wouldn't hurt though I guess since it will already be in that state.

natario1 replied:
Well it's not important then, as you wish

natario1 commented on library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:
```diff
@@ -1056,86 +1202,118 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
+                    it.cancel()
                 }
-            })
+                mContainer.postOnAnimation {
```
You don't need this post here, I think the animator already runs in an animation step.

markusressel replied:
ok

natario1 commented on library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:
```diff
@@ -1056,86 +1202,118 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
+                if (mClearAnimation) {
+                    it.cancel()
+                }
+                mContainer.postOnAnimation {
```
Same here, post is not needed. Nice approach

markusressel replied:
ok

natario1 commented on library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:
```diff
@@ -1147,60 +1325,63 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
      * @param allowOverScroll whether to overscroll
      */
-    private fun animateScaledPan(@ScaledPan deltaX: Float, @ScaledPan deltaY: Float,
+    private fun animateScaledPan(@AbsolutePan deltaX: Float, @ScaledPan deltaY: Float,
```
I think this was an error

markusressel replied:
jup good catch!

natario1 commented on library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:
```diff
@@ -1147,60 +1325,63 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
+                    it.cancel()
                 }
-            })
+                mContainer.postOnAnimation {
```
Same as before, post not needed

markusressel replied:
ok

markusressel committed:
added copy constructor to Points
removed unneeded post() calls
minimized constuctor calls on pan/scaledPan/mCurrentPanCorrection
removed unused method

markusressel committed:
review fixes

markusressel committed:
add pan to ZoomApi

natario1 commented on library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:
```diff
@@ -1151,56 +1329,57 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
+            panAnimator.addUpdateListener {
+                LOG.v("animateScaledPan:", "animationStep:", it.animatedFraction)
+                if (mClearAnimation) {
+                    it.cancel()
```
Sorry I forgot this, we need a return after these it.cancel() to be as it was.

markusressel replied:
right, I changed it in all 3 places.

markusressel committed:
review fix

natario1 reviewed (approved):
This is good IMO! It was a great addition, thanks.

markusressel committed:
Merge remote-tracking branch 'origin/master' into feature/#66_pan_while_zoom

# Conflicts:
#	library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt

markusressel committed:
merge and compile fix

markusressel commented:
Since you already approved this PR and I only fixed a small dev-merge mistake I will merge this PR now.
If any merge conflicts come up in #74 because of this and you are unsure what to do just let me know.
Cheers! :tada:
